mipengine/controller/celery_app.py

```diff
@@ -1,20 +1,178 @@
+from typing import Union
+
+import amqp
+import billiard
+import celery
+import kombu
 from celery import Celery
+from celery.canvas import Signature
+from celery.result import AsyncResult
 
 from mipengine.controller import config as controller_config
+from mipengine.singleton import Singleton
+
+
+class CeleryConnectionError(Exception):
+    def __init__(self, connection_address: str, error_details: str):
+        message = f"Connection Error: {connection_address=} {error_details=}"
+        super().__init__(message)
+
+
+class CeleryTaskTimeoutException(Exception):
+    def __init__(
+        self, timeout_type: str, connection_address: str, async_result: AsyncResult
+    ):
+        message = f"Timeout Exception: {timeout_type=} {connection_address=} while waiting for {async_result.task_id=}"
+        super().__init__(message)
+
+
+class CeleryWrapper:
+    def __init__(self, socket_addr: str):
+        self._socket_addr = socket_addr
+        self._celery_app = self._create_new_celery_app()
+
+    def _close(self):
+        self._celery_app.close()
+
+    # queue_task() is non-blocking, because apply_async() is non-blocking
+    def queue_task(self, task_signature: str, *args, **kwargs) -> AsyncResult:
+        try:
+            task_signature = self._celery_app.signature(task_signature)
+            async_result = task_signature.apply_async(args, kwargs)
+            return async_result
+        except (
+            kombu.exceptions.OperationalError,
+            amqp.exceptions.AccessRefused,
+            amqp.exceptions.NotAllowed,
+        ) as exc:
+            # TODO how are things supposed to be logged in here?
+            # get_backgroung_service_logger() makes no sense
+
+            ######################
+            print(f"(celery_app::CeleryWrapper::queue_task) {exc=}\n")
+
+            import traceback
+
+            tr = traceback.format_exc()
+            print(tr)
+            ######################
+
+            self._close()
+            self._celery_app = self._create_new_celery_app()
+
+            connection_error = CeleryConnectionError(
+                connection_address=self._socket_addr,
+                error_details=f"while queuing {task_signature=} with {args=} and {kwargs=}",
+            )
+
+            raise connection_error
+
+    # get_result() is blocking, because celery.result.AsyncResult.get() is blocking
+    def get_result(
+        self, async_result: AsyncResult, timeout: int
+    ) -> Union[str, dict, list]:
```

Comment:
If `celery.result.AsyncResult.get()` is called while the broker is down, exceptions are raised, which after being caught inside  the `get_result` method of `CeleryWrapper`, a new `Celery` object is instantiated, because the existing `Celery` object is now obsolete. 
If we implemented a separate AsyncResultWrapper class, as you say, we would not have access to the `CeleryWrapper` attribute `self._celery_app` to assign a new Celery object. Although, you could, for example, pass the `CeleryWrapper` object in the `get` method as a parameter, in order to be able to trigger the instantiation of a new `Celery` object if `get` method fails, this does not seem nice. Nevertheless, if you see another way, I do not see, feel free to outline it.

Reply:
I get your point. What about this?

```
class CeleryAsyncResult:
  def __init__(self, result: AsyncResult, celery_wrapper: CeleryWrapper):
    self.result= result
    self.celery_wrapper = celery_wrapper
  
  def get(timeout: int):
    return self.celery_wrapper.get_result(result, timeout)

class CeleryWrapper:
    def __init__(self, socket_addr: str):
        self._socket_addr = socket_addr
        self._celery_app = self._create_new_celery_app()

    def _close(self):
        self._celery_app.close()

    # queue_task() is non-blocking, because apply_async() is non-blocking
    def queue_task(self, task_signature: str, *args, **kwargs) -> AsyncResult:
        try:
            task_signature = self._celery_app.signature(task_signature)
            async_result = task_signature.apply_async(args, kwargs)
            return async_result
        except (
            kombu.exceptions.OperationalError,
            amqp.exceptions.AccessRefused,
            amqp.exceptions.NotAllowed,
        ) as exc:
            ...

    # get_result() is blocking, because celery.result.AsyncResult.get() is blocking
    def get_result(
        self, async_result: AsyncResult, timeout: int
    ) -> Union[str, dict, list]:
```

Reply:
this is almost what I described and mentioned it is not nice. It is also missing who is going to instantiate the `CeleryAsyncResult` object. Probably the `CeleryWrapper::queue_task` method would not return an `AsyncResult` but a `CeleryAsyncResult(result=async_result, celery_wrapper=self)` and this is very ugly.

Reply:
What you described was passing the `CeleryWrapper` object each time in the `get` method, which I agree is bad.

In the code snippet I showed, the `CeleryAsyncResult` is instantiated in the `queue_task` method.

So having a wrapper for the async result, where you could just `.get()` is ugly but having a specific `get_result` method for each task is pretty?
```
    # --------------- get_node_info task ---------------
    # NON-BLOCKING
    def queue_node_info_task(self, request_id: str) -> AsyncResult:
        celery_app = self._get_node_celery_app()
        task_signature = TASK_SIGNATURES["get_node_info"]
        try:
            async_result = celery_app.queue_task(
                task_signature=task_signature,
                request_id=request_id,
            )
            return async_result
        except CeleryConnectionError as exc:
            logger = ctrl_logger.get_request_logger(request_id=request_id)
            logger.error(exc)
            raise exc

    # BLOCKING
    def result_node_info_task(
        self, async_result: AsyncResult, request_id: str
    ) -> NodeInfo:
        celery_app = self._get_node_celery_app()
        try:
            result = celery_app.get_result(
                async_result=async_result,
                timeout=self._tasks_timeout,
                request_id=request_id,
            )
            return NodeInfo.parse_raw(result)
        except (CeleryTaskTimeoutException, ConnectionError) as exc:
            logger = ctrl_logger.get_request_logger(request_id=request_id)
            logger.error(exc)
            raise exc
```

or having to call the `get_result` on the `celery_app` instead of the actual result?
```
       async_result = celery_app.queue_task(
            task_signature=task_signature,
            table_name=table_name,
            table_schema_json=table_schema_json,
            monetdb_socket_address=original_db_url,
            request_id=request_id,
        )
        celery_app.get_result(
            async_result=async_result,
            timeout=self._tasks_timeout,
            request_id=request_id,
        )
```

I think it would remove much code duplication and make the code much cleaner, but it's up to you.

Reply:
the ugly part is the `return CeleryAsyncResult(async_result, self)` where you return an object that contains a reference to `self`.
The only change you d have if there was an implementation of `CeleryAsyncResult` it would be that instead of 
```
    async_result = celery_app.queue_task(
        task_signature=task_signature,
        table_name=table_name,
        table_schema_json=table_schema_json,
        monetdb_socket_address=original_db_url, 
        request_id=request_id, )

    result = celery_app.get_result(
        async_result=async_result,
        timeout=self._tasks_timeout,
        request_id=request_id, )
```
you d write
```
    async_result = celery_app.queue_task(
        task_signature=task_signature,
        table_name=table_name,
        table_schema_json=table_schema_json,
        monetdb_socket_address=original_db_url, 
        request_id=request_id, )

    result = async_result.get(
        timeout=self._tasks_timeout,
        request_id=request_id, )
```
Nevertheless, this is all caused by the need to instantiate a new `Celery` object if the broker goes down and up again, which seems to be a bug and will supposedly be fixed at some point, so we will be able to get rid of this workaround. But exactly because this is a workaround and it should be as contained as possible, the way I have implemented it, the workaround is only contained in one class, `CeleryWrapper`. In your implementation you split the workaround in two classes and moreover from the second class, `CeleryAsyncResult` you would have to be able to make changes on the `CeleryWrapper` object whenever a new `Celery` object would have to be instantiated. Both implementations have pros and cons, mine contains the workaround better, yours makes for better syntax on other parts of the code and would potentially need no changes when the workaround is not needed anymore. Nevertheless, I want to keep it like that for now and keep this conversation as reference when the bug is fixed and the workaround is not needed anymore.